Fix custominstallerbuilder/issue#16

lukpueh opened the pull request:
Fixes [custominstallerbuilder/issue16](https://github.com/SeattleTestbed/custominstallerbuilder/issues/16)
1. Adds big red warning to all web pages when `DEBUG` option is set to true.
2. Adds option to chose whether the Clearinghouse should be able to communicate with a Custom Installer Builder via XMLRPC if the Custom Installer Builder uses _self-signed certificates_.
   Option defaults to the same value as `DEBUG` option.

There are more parts in the Clearinghouse code that use XMLRPC but were not changed because they either don't use SSL or aren't probably used at all:
- [lockserver](https://github.com/SeattleTestbed/clearinghouse/blob/master/common/api/lockserver.py#L351)
- [backend](https://github.com/SeattleTestbed/clearinghouse/blob/master/common/api/backend.py#L63)
- [lock_contention](https://github.com/SeattleTestbed/clearinghouse/blob/master/lockserver/tests/integration/lock_contention.py)
- [ut_integration_lock_contention.py](https://github.com/SeattleTestbed/clearinghouse/blob/master/lockserver/tests/integration/ut_integration_lock_contention.py)
- [example1](https://github.com/SeattleTestbed/clearinghouse/blob/master/xmlrpc_clients/examples/example1.py#L27)

lukpueh commented on website/html/views.py:
```diff
@@ -22,6 +22,8 @@
 import shutil
 import subprocess
 import xmlrpclib
+import re
```
This import slipped in from a previous commit (wrong fork, wrong branch, had to cherry pick, resolve conflicts, ...).
Anyways, I actually think it already slipped into the original [commit](https://github.com/kaushik1091/clearinghouse/pull/1/commits/74fad83406fa8a06a3384ba0f52622fc4434eb32) by accident. Because I can't see the `re` module being used.

JustinCappos replied:
So should it be removed?

vladimir-v-diaz commented on website/settings.py:
```diff
@@ -51,6 +51,13 @@
+# This setting provides to chose whether to accept self-signed certificates
+# (e.g. in DEBUG mode) by using an unverified ssl context.
+# c.f. https://github.com/SeattleTestbed/custominstallerbuilder/issues/16
+UNVERIFIED_SSL_CONTEXT = DEBUG
```
I suppose `UNVERIFIED_SSL_CONTEXT` can be set to True independent of DEBUG.  In that case, I can imagine someone doing so and failing to notice that self-signed certificates are used in production (because they forgot it was set to True).

Consider having some type of warning for when self-signed certificates are accepted by the server.

lukpueh replied:
I think that that's not necessary, since the variable is currently tied to the debug mode and shouldn't really be set independently anyway.

vladimir-v-diaz replied:
Why have a separate UNVERIFIED_SSL_CONTEXT setting?  Perhaps you should make it so that DEBUG results in the ability to use unverified SSL contexts.

Someone might try to modify if you don't at least explain that it shouldn't be independently modified.

lukpueh replied:
I proposed the separate setting mostly for transparency. We never instruct a *ch operator* to set this to a different value than `DEBUG` and I can't see why a *ch operator* would want to do this, or would do this by accident.

vladimir-v-diaz commented on website/html/views.py:
```diff
@@ -1107,7 +1108,17 @@ def _build_installer(username, platform):
+    if settings.UNVERIFIED_SSL_CONTEXT:
+      xmlrpc_proxy = xmlrpclib.ServerProxy(
+        settings.SEATTLECLEARINGHOUSE_INSTALLER_BUILDER_XMLRPC, 
+        context=ssl._create_unverified_context())
```
Isn't this function only available in Python version 2.7+?  What if a user is using version <2.7?  Should there be checks for Python versions in the code, or some kind of disclaimer?

aaaaalbert replied:
Python 2.6 is plainly *unsupported* for running a clearinghouse as of the docs change.

lukpueh commented on website/html/views.py:
```diff
@@ -1107,7 +1108,17 @@ def _build_installer(username, platform):
+    if settings.UNVERIFIED_SSL_CONTEXT:
+      xmlrpc_proxy = xmlrpclib.ServerProxy(
+        settings.SEATTLECLEARINGHOUSE_INSTALLER_BUILDER_XMLRPC, 
+        context=ssl._create_unverified_context())
```
AFAIK we agreed that we don't care for Python <2.7 to run clearinghouse (c.f. discussion in SeattleTestbed/custominstallerbuilder#16).

As for the disclaimer, our main instruction to run the clearinghouse already [suggests to use Python 2.7](https://github.com/SeattleTestbed/docs/blob/master/Operating/Clearinghouse/Installation.md)

lukpueh commented:
Thanks for reviewing @vladimir-v-diaz and @JustinCappos! I have addressed your comments and also fixed a small but important bug in my code. Let me know if we can merge.

vladimir-v-diaz commented on website/settings.py:
```diff
@@ -51,6 +51,13 @@
+# This setting provides to chose whether to accept self-signed certificates
+# (e.g. in DEBUG mode) by using an unverified ssl context.
+# c.f. https://github.com/SeattleTestbed/custominstallerbuilder/issues/16
+UNVERIFIED_SSL_CONTEXT = DEBUG
```
I understand, but the hypothetical situation described is for a user that sets UNVERIFIED_SSL_CONTEXT to True independent of DEBUG.  In other words, someone might try to set DEBUG to False, but UNVERIFIED_SSL_CONTEXT to True.

Are you saying that users should not attempt to modify UNVERIFIED_SSL_CONTEXT, and only modify DEBUG?  If so, that should documented somewhere.

vladimir-v-diaz commented:
Feel free to merge!
